### PR TITLE
[CARBONDATA-701]Fix memory leak issue for new flow dataloading

### PR DESCRIPTION
**Pull request by QiangCai**
1. release RecordReader resources one by one , not release all resouces at the end of sort step.
2. reduce writer buffer size(50000 => 16KB)
3. reduce prefetch size to reduce memory taken in final merger queue. 
4. other modifcations
     a) refactory CarbonRowBatch, use Array instead of ArrayList
     b) fix a bug in SortDataRow.java, the bug is the result maybe is wrong if total row number is bigger than 120000. 

Heap size: old code VS new code
old code:
10million rows: 1.8G
old_10million
100million rows: 3G max size
old_100million
100million rows with prefetch size 1000: 3G max size
old_100million_pretech
100million rows with kettle flow: 2.6GB
100million_kettle
200million rows: 3G max size, loading can't be finished, I killed it.
old_200million

new code:
10million rows: 1.5GB
new_10million
100million rows: 1.5GB
new_100million
200million rows: 1.5GB
new_200million
500million rows: 2GB
new_500million

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/877/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/879/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/889/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/892/

File: processing/src/main/java/org/apache/carbondata/processing/newflow/row/CarbonRowBatch.java [hunk 1]
```diff
@@ -17,27 +17,40 @@
...
-import java.util.List;
 
 /**
  * Batch of rows.
  */
-public class CarbonRowBatch {
+public class CarbonRowBatch implements Iterator<CarbonRow> {
 
-  private List<CarbonRow> rowBatch = new ArrayList<>();
+  private CarbonRow[] rowBatch;
```
Comment (chenliang613) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/row/CarbonRowBatch.java]: Here, Why changes ArrayList to Array?

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 2]
```diff
@@ -641,7 +641,7 @@
   /**
    * CARBON_PREFETCH_BUFFERSIZE
    */
-  public static final int CARBON_PREFETCH_BUFFERSIZE = 20000;
+  public static final int CARBON_PREFETCH_BUFFERSIZE = 1000;
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: Please take this from carbon properties file, and keep default as 1000

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/ParallelReadMergeSorterImpl.java [hunk 3]
```diff
@@ -86,11 +88,10 @@ public void initialize(SortParameters sortParameters) {
             sortParameters.getNoDictionaryDimnesionColumn(), sortParameters.isUseKettle());
   }
 
-  @Override
-  public Iterator<CarbonRowBatch>[] sort(Iterator<CarbonRowBatch>[] iterators)
+  public void prepare(Iterator<CarbonRowBatch>[] iterators)
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/ParallelReadMergeSorterImpl.java]: Better change name to initialize

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java [hunk 4]
```diff
@@ -116,8 +116,15 @@ public IntermediateFileMerger(SortParameters mergerParameters, File[] intermedia
           writeDataTofile(next());
         }
       } else {
+        int i = 0;
         while (hasNext()) {
+          i++;
           writeDataTofileWithOutKettle(next());
+          if (i % 10000 == 0) {
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java]: it is unnecessary calculation we are adding in it. Already the buffer size is configurable so once buffer size reaches buffered out stream can flush it down.

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java [hunk 5]
```diff
@@ -375,6 +376,9 @@ private void writeDataWithOutKettle(Object[][] recordHolderList, int entryCountL
             stream.write((byte) 0);
           }
         }
+        if (i % 10000 == 9999) {
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: it is unnecessary calculation we are adding in it. Already the buffer size is configurable so once buffer size reaches buffered out stream can flush it down.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/900/

File: processing/src/main/java/org/apache/carbondata/processing/newflow/row/CarbonRowBatch.java (hunk 1 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/row/CarbonRowBatch.java]: The default capacity of ArrayList is 10.  Before adding a element, need to ensure capacity, and grow capacity if needed.

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/ParallelReadMergeSorterImpl.java (hunk 3 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/impl/ParallelReadMergeSorterImpl.java]: Yes, but the method initialize is exists.

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java (hunk 4 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java]: ok, I will test a better value of defualt buffer size.

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java (hunk 5 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/SortDataRows.java]: fixed

File: processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java (hunk 4 above)
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/sortandgroupby/sortdata/IntermediateFileMerger.java]: fixed

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java (hunk 2 above)
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java]: fixed

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/908/

File: processing/src/main/java/org/apache/carbondata/processing/newflow/sort/Sorter.java [hunk 6]
```diff
@@ -39,11 +39,13 @@
...
    *
-   * @param iterators array of iterators to read data.
    * @return
    * @throws CarbonDataLoadingException
    */
-  Iterator<CarbonRowBatch>[] sort(Iterator<CarbonRowBatch>[] iterators)
+  Iterator<CarbonRowBatch>[] sort()
+      throws CarbonDataLoadingException;
+
+  void prepare(Iterator<CarbonRowBatch>[] iterators)
```
Comment (ravipesala) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/Sorter.java]: Already previous steps file resources are closed where and when the use of it is over. I don't think this method is required.

Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/Sorter.java]: Better to invoke child.close() before final merger

Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/newflow/sort/Sorter.java]: fixed

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/911/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/912/

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**